KOGITO-5217:Add Data Index connection with runtimes endpoints: Abort,…

nmirasch opened the pull request:
… Retry, pInstanceDiagram, pInstanceNodes

Jira related : https://issues.redhat.com/browse/KOGITO-5217

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Pull Request</b>  
  Please add comment: <b>Jenkins retest this</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

</details>

nmirasch commented:
Jenkins retest this

cristianonicolai commented on data-index/data-index-service/data-index-service-common/src/main/resources/schema.graphqls:
```diff
@@ -11,6 +12,13 @@ type Query {
 }
 
+type Mutation {
+    ProcessInstanceAbort(processInstanceId: String): String
```
```suggestion
    ProcessInstanceAbort(id: String): String
```

nmirasch replied:
changed!

cristianonicolai commented on data-index/data-index-service/data-index-service-common/src/main/resources/schema.graphqls:
```diff
@@ -11,6 +12,13 @@ type Query {
 
+type Mutation {
+    ProcessInstanceAbort(processInstanceId: String): String
+    ProcessInstanceRetry(processInstanceId: String): String
```
```suggestion
    ProcessInstanceRetry(id: String): String
```

nmirasch replied:
changed!

nmirasch commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/resources/DataIndexPostgreSqlResource.java:
```diff
@@ -83,6 +86,7 @@ public int getMappedPort() {
     }
 
     public Map<String, String> getProperties() {
+        //properties.put(DATA_INDEX_BLOCKING, "true");
```
removed!

nmirasch commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/testcontainers/DataIndexPostgreSqlContainer.java:
```diff
@@ -31,6 +31,7 @@ public void setPostgreSqlURL(String postgreSqlURL, String username, String passw
         addEnv("QUARKUS_DATASOURCE_JDBC_URL", postgreSqlURL);
         addEnv("QUARKUS_DATASOURCE_USERNAME", username);
         addEnv("QUARKUS_DATASOURCE_PASSWORD", password);
+        addEnv(DATA_INDEX_BLOCKING, "true");
```
yes!

nmirasch commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-quarkus/src/test/java/org/kie/kogito/index/quarkus/DataIndexMongoDBQuarkusTestResource.java:
```diff
@@ -34,6 +35,7 @@ public DataIndexMongoDBQuarkusTestResource() {
     protected Map<String, String> getProperties() {
         Map<String, String> properties = new HashMap<>();
         properties.put(KOGITO_DATA_INDEX_SERVICE_URL, "http://localhost:" + getTestResource().getMappedPort());
+        properties.put(KOGITO_DATA_INDEX_HTTP_URL, "http://localhost:" + getTestResource().getMappedPort());
```
unified!

nmirasch commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-quarkus/src/test/java/org/kie/kogito/index/quarkus/KogitoServiceRandomPortQuarkusTestResource.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    protected Map<String, String> getProperties() {
+        System.setProperty(QUARKUS_SERVICE_HTTP_PORT, String.valueOf(getTestResource().getMappedPort()));
```
yes!

nmirasch commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/AbstractProcessDataIndexIT.java:
```diff
@@ -209,6 +215,94 @@ public void testProcessInstanceEvents() {
+        getProcessInstanceById(pId2, "ACTIVE");
+
+        await().untilAsserted(() -> given().spec(dataIndexSpec()).contentType(ContentType.JSON)
+                .body("{ \"query\" : \"mutation{ GetProcessInstanceDiagram( processInstanceId: \\\"" + pId2 + "\\\")}\"}")
```
Done, even I was thinking in remove the query GetProcessInstanceNode.

nmirasch commented:
Jenkins retest this

nmirasch commented:
Jenkins retest this

cristianonicolai commented on data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/GraphQLSchemaManager.java:
```diff
@@ -102,12 +109,21 @@ private GraphQLSchema createSchema() {
                     builder.dataFetcher("ProcessInstances", this::getProcessInstancesValues);
                     builder.dataFetcher("UserTaskInstances", this::getUserTaskInstancesValues);
                     builder.dataFetcher("Jobs", this::getJobsValues);
+                    builder.dataFetcher("GetProcessInstanceNodes", this::getProcessInstanceNodes);
```
to be removed

nmirasch replied:
done

cristianonicolai commented on data-index/data-index-service/data-index-service-common/src/main/resources/schema.graphqls:
```diff
@@ -92,6 +102,14 @@ type NodeInstance {
     nodeId: String!
 }
 
+type AvailableNode {
```
NodeDefinition

cristianonicolai replied:
NodeInstance extends NodeDefinition

nmirasch replied:
Done!

cristianonicolai commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/testcontainers/DataIndexPostgreSqlContainer.java:
```diff
@@ -20,7 +20,7 @@
  */
 public class DataIndexPostgreSqlContainer extends AbstractDataIndexContainer {
-
+    public static final String DATA_INDEX_BLOCKING = "kogito.data-index.blocking";
```
probably not needed?

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

nmirasch commented:
jenkins please retest this

cristianonicolai commented on data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/DataIndexApiImpl.java:
```diff
@@ -0,0 +1,179 @@
+import static java.lang.String.format;
+
+@ApplicationScoped
+public class DataIndexApiImpl implements DataIndexApi {
```
perhaps rename to KogitoRuntimeClient

cristianonicolai commented on data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/DataIndexApiImpl.java:
```diff
@@ -0,0 +1,179 @@
+        if (new ProcessInstanceInErrorPredicate().test(processInstance)) {
+            return sendPostClientRequest(getWebClient(serviceURL), requestURI, "SKIP ProcessInstance with id: " + processInstance.getId());
+        }
+        CompletableFuture<String> future = new CompletableFuture<>();
```
```suggestion
        CompletableFuture.failedFuture()
```

cristianonicolai commented on data-index/data-index-service/data-index-service-common/src/main/resources/schema.graphqls:
```diff
@@ -92,6 +101,14 @@ type NodeInstance {
     nodeId: String!
 }
 
+type Node {
```
NodeInstance can extend Node

nmirasch replied:
@cristianonicolai It seems that in GraphQL an input type can't inherit from another type or interface. Even the a type can be extended with more fields(the extends), but the result is not a new type.  Maybe we can create a field inside NodeInstance that contains the Node?

cristianonicolai replied:
lets leave as its then

cristianonicolai commented on data-index/data-index-storage/data-index-storage-api/src/main/java/org/kie/kogito/index/model/Node.java:
```diff
@@ -0,0 +1,101 @@
+public class Node {
+
+    private String id;
+    @JsonProperty("name")
+    private String name;
+    @JsonProperty("type")
```
as the property name is the same as the variable name, you don't need to annotate.

nmirasch commented:
jenkins please retest this

nmirasch commented:
jenkins please retest this

nmirasch commented:
@cristianonicolai @radtriste  
Added the mutations/queries to data index to be able to execute runtime operations. And also the integration test to see it working with the different persistence options. In order to avoid the time out in the integration test I have removed the process instance diagram query. Could you take a look on this? thanks!!

cristianonicolai commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/AbstractProcessDataIndexIT.java:
```diff
@@ -209,6 +216,65 @@ public void testProcessInstanceEvents() {
+                        .statusCode(200)
+                        .body("data.ProcessInstances.nodeDefinitions.size()", is(1)));
+
+        await()
+                .atMost(TIMEOUT)
+                .untilAsserted(() -> given().spec(dataIndexSpec()).contentType(ContentType.JSON)
+                        .body("{ \"query\" : \"mutation{ ProcessInstanceAbort( id: \\\"" + pId2 + "\\\")}\"}")
+                        .when().post("/graphql")
+                        .then().statusCode(200));
+
+        await()
+                .atMost(TIMEOUT)
+                .untilAsserted(() -> given().spec(dataIndexSpec()).contentType(ContentType.JSON)
+                        .body("{ \"query\" : \"mutation{ ProcessInstanceRetry( id: \\\"" + pId2 + "\\\")}\"}")
+                        .when().post("/graphql")
+                        .then()
+                        .statusCode(200));
+        await()
+                .atMost(TIMEOUT)
+                .untilAsserted(() -> given().spec(dataIndexSpec()).contentType(ContentType.JSON)
+                        .body("{ \"query\" : \"mutation{ ProcessInstanceSkip( id: \\\"" + pId2 + "\\\")}\"}")
+                        .when().post("/graphql")
+                        .then()
+                        .statusCode(200));
```
@nmirasch mutation queries do not need to await as they are meant to be executed only once. Is like a PUT, or POST.

nmirasch replied:
removed await from mutations request.

cristianonicolai commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/AbstractProcessDataIndexIT.java:
```diff
@@ -209,6 +216,65 @@ public void testProcessInstanceEvents() {
+        await()
+                .atMost(TIMEOUT)
+                .untilAsserted(() -> given().spec(dataIndexSpec()).contentType(ContentType.JSON)
+                        .body("{ \"query\" : \"query { ProcessInstances (where: { id: {equal: \\\"" + pId2 + "\\\"}}) {nodeDefinitions {id}} }\" }")
```
```suggestion
                        .body("{ \"query\" : \"ProcessInstances (where: { id: {equal: \\\"" + pId2 + "\\\"}}) {nodeDefinitions {id}} }\" }")
```

nmirasch replied:
done!

cristianonicolai commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/AbstractProcessDataIndexIT.java:
```diff
@@ -209,6 +216,65 @@ public void testProcessInstanceEvents() {
+        await()
+                .atMost(TIMEOUT)
+                .untilAsserted(() -> given().spec(dataIndexSpec()).contentType(ContentType.JSON)
+                        .body("{ \"query\" : \"mutation{ ProcessInstanceAbort( id: \\\"" + pId2 + "\\\")}\"}")
```
would this work? I feel that query should not be used in this case
```suggestion
                        .body("{ \"mutation\" : \"ProcessInstanceAbort( id: \\\"" + pId2 + "\\\")}\"}")
```

nmirasch replied:
AFAIK this only work if we pass the operation in field named 'query'.. https://graphql.org/learn/serving-over-http/#post-request

cristianonicolai commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-quarkus/pom.xml:
```diff
@@ -12,11 +12,29 @@
+    <dependencies>
+      <dependency>
+        <groupId>org.kie.kogito</groupId>
+        <artifactId>integration-tests-common</artifactId>
```
maybe add this into kogito-apps-bom?

nmirasch replied:
I've removed the dependency, I just brought that following the jobs integration test, but it seems it's not needed.

cristianonicolai commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-springboot/src/test/java/org/kie/kogito/index/ProcessDataIndexInfinispanIT.java:
```diff
@@ -28,8 +29,8 @@
 
-@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, classes = { KogitoApplication.class })
-@ContextConfiguration(initializers = DataIndexInfinispanSpringTestResource.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT, classes = { KogitoApplication.class })
```
shouldnt we keep the random port?

nmirasch replied:
yes,makes sense, restored!

cristianonicolai commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-springboot/src/test/java/org/kie/kogito/index/ProcessDataIndexPostgreSqlIT.java:
```diff
@@ -28,8 +29,8 @@
 
-@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, classes = { KogitoApplication.class })
-@ContextConfiguration(initializers = DataIndexPostgreSqlSpringTestResource.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT, classes = { KogitoApplication.class })
```
same here, preferably to use random port

cristianonicolai commented on apps-integration-tests/integration-tests-data-index-service/pom.xml:
```diff
@@ -46,4 +46,33 @@
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-failsafe-plugin</artifactId>
+        <!-- overriding parent configuration because quarkus.http.test-port should not be set -->
```
what was the issue with that? If we could avoid this override just because of that would be good

nmirasch replied:
same as above, this is brought from the jobs integration test, but it seems it's not needed. Removed!

cristianonicolai commented on data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/GraphQLSchemaManager.java:
```diff
@@ -134,6 +148,31 @@ private GraphQLSchema createSchema() {
+        return dataIndexApiExecutor.skipProcessInstance(getServiceUrl(processInstance.getEndpoint(), processInstance.getProcessId()), processInstance);
+    }
+
+    public CompletableFuture getProcessInstanceDiagram(DataFetchingEnvironment env) {
```
```suggestion
    public CompletableFuture<String> getProcessInstanceDiagram(DataFetchingEnvironment env) {
```

nmirasch replied:
added!

cristianonicolai commented on data-index/data-index-service/data-index-service-common/src/test/java/org/kie/kogito/index/graphql/query/AbstractGraphQLRuntimesQueriesIT.java:
```diff
@@ -0,0 +1,132 @@
+        KogitoProcessCloudEvent startEvent = getProcessCloudEvent(processId, processInstanceId, ACTIVE, null, null, null);
+        indexProcessCloudEvent(startEvent);
+
+        checkOkResponse("{ \"query\" : \"mutation{ ProcessInstanceAbort ( id: \\\"" + processInstanceId + "\\\")}\"}");
```
```suggestion
        checkOkResponse("{ \"mutation\" : \"ProcessInstanceAbort ( id: \\\"" + processInstanceId + "\\\")}\"}");
```

nmirasch replied:
same here, we need to pass the operation in a 'query' field

cristianonicolai commented:
Jenkins rerun apps tests

nmirasch commented:
Jenkins retest this

nmirasch commented:
jenkins please retest this

nmirasch commented:
jenkins please retest this

radtriste reviewed (approved):
2 minor things

nmirasch commented on apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/utils/SocketUtils.java:
```diff
@@ -0,0 +1,64 @@
+
+import javax.net.ServerSocketFactory;
+
+public final class SocketUtils {
```
Yes, I agree, that is the right place. Let me try to do that movement as part of next PR as part of: https://issues.redhat.com/browse/KOGITO-5553

nmirasch commented:
jenkins please retest this

nmirasch commented:
jenkins please retest this

nmirasch commented:
jenkins please retest this
